### PR TITLE
[Frequency analyzer part 3] Convert the raw_df to long format

### PR DESCRIPTION
**Pull request by huangxichen1**
This is to provide data for generating barplot for comparing frequency distributions.

### CODE REVIEW

File: src/evaluations/tests/analyzer_test.py [hunk 1]
```diff
@@ -254,5 +256,45 @@ def test_get_analysis_results(self):
...
+  def test_convert_raw_df_to_long_format(self):
+    df = pd.DataFrame({
+        analyzer.SKETCH_ESTIMATOR_NAME: ['some_sketch'] * 4,
+        analyzer.SCENARIO_NAME: ['some_scenario'] * 4,
+        simulator.RUN_INDEX: [0, 0, 1, 1],
+        simulator.NUM_SETS: [1, 2, 1, 2],
+        simulator.TRUE_CARDINALITY_BASENAME + '1': [10, 20, 10, 20],
+        simulator.TRUE_CARDINALITY_BASENAME + '2': [5, 10, 5, 10],
+        simulator.ESTIMATED_CARDINALITY_BASENAME + '1': [11, 21, 12, 22],
+        simulator.ESTIMATED_CARDINALITY_BASENAME + '2': [4, 9, 3, 8],
```
Comment (my2457) [inline src/evaluations/tests/analyzer_test.py:270]: here '1', '2' refer to the frequency cap, correct?

**Review by my2457 — APPROVED** [review]
lgtm

**Review by shengma-1 — APPROVED** [review]

File: src/evaluations/tests/analyzer_test.py (hunk 1 above)
Comment (huangxichen1) [reply src/evaluations/tests/analyzer_test.py:270]: They refer to the n+ reach. For example, '1' refers to the 1+ reach, and '2' refers to the 2+ reach (i.e., number of ids who are seen at least twice in the campaign). And the maximum `n` can be controlled by the config.

File: src/evaluations/analyzer.py [hunk 2]
```diff
@@ -278,6 +282,60 @@ def __init__(self, out_dir, evaluation_directory, evaluation_run_name,
     super().__init__(out_dir, evaluation_directory, evaluation_run_name,
                      evaluation_name, estimable_criteria_list, plot_params)
 
+  def convert_raw_df_to_long_format(self):
+    """Convert the raw DataFrame to a long format.
```
Comment (kungfucraig) [inline src/evaluations/analyzer.py:286]: Should probably describe what the long format is somewhere in the doc string.

File: src/evaluations/analyzer.py [hunk 3]
```diff
@@ -278,6 +282,60 @@ def __init__(self, out_dir, evaluation_directory, evaluation_run_name,
...
+
+    Returns:
+      A pd.DataFrame. It is a long format of self.raw_df, which contains
+      columns of
```
Comment (kungfucraig) [inline src/evaluations/analyzer.py:290]: This looks incomplete.

File: src/evaluations/analyzer.py [hunk 4]
```diff
@@ -278,6 +282,60 @@ def __init__(self, out_dir, evaluation_directory, evaluation_run_name,
...
+
+    df_long = self.raw_df.melt(
+        id_vars=[SKETCH_ESTIMATOR_NAME, SCENARIO_NAME, simulator.RUN_INDEX,
+                 simulator.NUM_SETS],
+        value_vars=value_vars,
+        var_name=CARDINALITY_SOURCE,
+        value_name=CARDINALITY_VALUE,
+    )
+
+    def _split_source_and_frequency(source_freq):
```
Comment (kungfucraig) [inline src/evaluations/analyzer.py:308]: Would recommend making this module level. At the very least it should not be interspersed in the middle of a function.

File: src/evaluations/analyzer.py [hunk 5]
```diff
@@ -278,6 +282,60 @@ def __init__(self, out_dir, evaluation_directory, evaluation_run_name,
...
+      """Split the cardinality_frequency string into a named pd.Series.
+
+      For example, true_cardinality_2 will be split into
+      pd.Series({CARDINALITY_SOURCE: 'true_cardinality', FREQUENCY_LEVEL: 2}).
+
+      Args:
+        source_freq: a string in the format of estimated_cardinality_X or
+          true_cardinality_X, where X is an integer.
+      Returns:
+        A named pd.Series with .
```
Comment (kungfucraig) [inline src/evaluations/analyzer.py:318]: This looks incomplete.

**Review by kungfucraig — CHANGES_REQUESTED** [review]

File: src/evaluations/analyzer.py (hunk 2 above)
Comment (huangxichen1) [reply src/evaluations/analyzer.py]: Done.

File: src/evaluations/analyzer.py (hunk 3 above)
Comment (huangxichen1) [reply src/evaluations/analyzer.py]: Done.

File: src/evaluations/analyzer.py (hunk 4 above)
Comment (huangxichen1) [reply src/evaluations/analyzer.py]: Done. Move it to be a (private) classmethod.

File: src/evaluations/analyzer.py (hunk 5 above)
Comment (huangxichen1) [reply src/evaluations/analyzer.py]: Done.

File: src/evaluations/analyzer.py [hunk 6]
```diff
@@ -278,6 +282,81 @@ def __init__(self, out_dir, evaluation_directory, evaluation_run_name,
...
+    r"""Convert the raw DataFrame to a long format.
+
+    The raw_df is in a wide format, i.e, it has multiple columns to represent
+    the cardinality of different frequency levels. For example, the columns
+    look like:
+    num_sets,estimated_cardinality_1,estimated_cardinality_2,
+    true_cardinality_1,true_cardinality_2\n
+    1,11,4,10,5\n
+    2,22,8,20,10
+    The wide format is difficult to be passed to the seaborn plotting functions.
```
Comment (kungfucraig) [inline src/evaluations/analyzer.py:314]: ... difficult to pass ...

**Review by kungfucraig — APPROVED** [review]

**Pull request merged**